specs/withdrawal-proof.md

```diff
@@ -0,0 +1,73 @@
+
+```
+withdrawal_0 = [withdrawal_index_0, validator_index_0, address_0, amount_0]
+withdrawal_1 = [withdrawal_index_1, validator_index_1, address_1, amount_1]
+withdrawals = [withdrawal_0, withdrawal_1]
+
+execution_payload_body_rlp = RLP([transactions, [], withdrawals])
+```
+
+
+## Circuit behavior
+
+For every withdrawal defined as the parameters `(withdrawal_index, validator_index, address, amount)` and the circuit verifies the followings:
+
+1. `withdrawalsData: bytes = rlp([withdrawal_index, validator_index, address, amount])`
+2. `withdrawalsRoot: word = mpt(withdrawalsData)`
+3. `withdrawal_index`, `validator_index` and `amount` are all `uint64` values.
+4. `amount_wei = amount * 1e9` and increases validator's balance by `amount_wei`
+
+- The rlp encoding of withdrawal parameters will be done using a custom rlp encoding gadget,  isolated from the rlp encoding used by the MPT circuit.
+- The MPT root verification will be done in MPT circuit; the withdrawal circuit do a lookup to the MPT table.
+
+From this information the circuit builds the WithdrawalTable:
+
+Where:
+
+- Address = validator's address
+- Amount = a nonzero amount of ether given in Gwei (1e9 wei)
+- MPT root = an incremental MPT root
+
+| 0 Withdrawal ID | 1 Validator ID | 2 Address      | 3 Amount      | 4 MPT root     |
+| -----------     | -------------  | -------------- | ------------- | -------------- |
+| $WithdrawalID   | $ValidatorID   | $value{Lo,Hi}  | $value{Lo,Hi} | $value{Lo,Hi}  |
+
+There are some constraints on the shape of the table like:
+
+- `WithdrawalID` is increased monotonically and sequentially for each withdrawal. Which means it won't be starting from 0 in most of cases.
+- MPT root is used to lookup MPT table.
+
+Except MPT root, the withdrawal table is built by public input circuit, the public input circuit would validate all rows of the table are properly built. MPT root is part of witness and the value is verified by MPT circuit. Since the table construction is validated outside of the circuit, there's no need to verify the same constraints inside of the circuit. 
+
+### Withdrawal Trie
+
+For each withdrawal, the withdrawal circuit also must prepare the key and value used to build the withdrawal trie.  These keys and values are used in lookups to the MPT table in order to verify that a tree built with the key-values corresponding to the withdrawals has the root value `withdrawalsRoot`.
+
+> By doing lookups to the MPT table, we prove that when we start with an empty MPT, and do a chain of key-value insertions corresponding to each transaction, we reach a Trie with root value `withdrawalsRoot`.
+
+Each MPT update uses the following parameters:
+
+- Key = `rlp(withdrawal_index)`
+- Value = `rlp([withdrawal_index, validator_index, address, amount])`
+- ValuePrev = `rlp([withdrawal_index_prev, validator_index_prev, address_prev, amount_prev])`
```

Comment:
Since it's a stack trie, I think `ValuePrev` should always be empty (non-exist)?

Reply:
fixed in [3444436](https://github.com/privacy-scaling-explorations/zkevm-specs/pull/487/commits/34444362e060d2f85525c48a8557d79a6017c836)